### ISSUE TITLE
Incorrect Xbox 360 Wireless Receiver Input Mapping on Linux

### ISSUE THREAD

**Issue by skison** [godotengine/godot#35575]
Using Godot v3.1.2-stable, with Ubuntu 18.04.3 LTS

Alright, so I have an Xbox One S Bluetooth controller, that when I connect it to my machine directly, Godot is able to correctly interpret the inputs. However when I use a device from 8bitdo, their "Wireless Bluetooth Adapter", after I connect my controller to it the input mapping is all wrong. This doesn't appear to be a problem on my Windows machine, only while using Linux.

When I was troubleshooting, I used the site html5gamepad.com to help identify the issue, and I noticed that it sees the 8bitdo adapter in this case as "045e-0719-Xbox 360 Wireless Receiver", whereas with the same controller directly connected through Bluetooth it is seen as "045e-02e0-Xbox Wireless Controller". On that site, both controller types and their mappings appear to be interpreted correctly.

Onto the specifics of the mapping issues in Godot, I used the Demo project "Joypads Demo/Tool" (found here: https://github.com/godotengine/godot-demo-projects/tree/3.1-c34a2b4/misc/joypads) to help me visualize the issues:
- Left joystick left,right,up,down all map correctly
- Right joystick up maps to right joystick left
- Right joystick right maps to left trigger
- Right joystick down maps to right joystick right
- Right joystick left maps to the same axis as right joystick right, though it doesn't show up as left trigger
- L3 (left joystick click) maps to d-pad left
- R3 (right joystick click) maps to d-pad right
- D-pad up and down don't appear to map to anything
- D-pad left maps to d-pad up
- D-pad right maps to d-pad down
- X, Y, A, B all map correctly
- Left and right bumpers map correctly
- Left trigger maps to select
- Right trigger maps to start
- Select maps to home
- Start maps to L3
- Home maps to R3

If it helps, I used a simple function:
```
func _input(event):
	print(event.as_text())
```
to print out the strings for the incorrectly mapped buttons, here are some of the results for that:
- Right joystick up: `InputEventJoypadMotion : axis=2, axis_value=-0.117191` (axis value from 0 to -1)
- Right joystick right: `InputEventJoypadMotion : axis=6, axis_value=0.5` (axis value from .5 to 1)
- Right joystick down: `InputEventJoypadMotion : axis=2, axis_value=0.093722` (axis value from 0 to 1)
- Right joystick left: `InputEventJoypadMotion : axis=6, axis_value=0.363277` (axis value from .5 to 0)
- L3: `InputEventJoypadButton : button_index=14, pressed=true, pressure=0`
- R3: `InputEventJoypadButton : button_index=15, pressed=true, pressure=0`
- No events for d-pad up or down
- D-pad left: `InputEventJoypadButton : button_index=12, pressed=true, pressure=0`
- D-pad right: `InputEventJoypadButton : button_index=13, pressed=false, pressure=0`
- Left trigger: `InputEventJoypadButton : button_index=10, pressed=false, pressure=0`
- Right trigger: `InputEventJoypadButton : button_index=11, pressed=true, pressure=0`
- Select: `InputEventJoypadButton : button_index=16, pressed=false, pressure=0`
- Start: `InputEventJoypadButton : button_index=8, pressed=true, pressure=0`
- Home: `InputEventJoypadButton : button_index=9, pressed=true, pressure=0`

To reproduce this issue, you would presumably need to get a device that Linux interprets as an "Xbox Wireless Receiver" equivalent to the one by 8bitdo, then start up any project in Godot with input mapping where the inputs can be visualized, then check the inputs I listed above to see the incorrect mappings (I noticed this issue persists in exported projects as well).

**Comment by akien-mga** [comment]
Could you try Godot 3.2 RC 3? I might have fixed this a few weeks ago.

**Comment by skison** [comment]
Unfortunately the same issue appears in v3.2-rc3.

**Comment by akien-mga** [comment]
Can you check if mappings for your controller are properly included in https://github.com/gabomdq/SDL_GameControllerDB ? See https://github.com/gabomdq/SDL_GameControllerDB#mapping-tools for tools that can let you generate the mapping.

If it's missing there, you can make a PR to that repo and we'll get the updated mapping next time I sync. If it's already included and working, it might be discarded in Godot due to #8368.

**Comment by skison** [comment]
I did find the same GUID/controller type listed in gamecontrollerdb.txt, but the mapping is different. What would be the best way to proceed?

**Comment by akien-mga** [comment]
You mean that you get a different mapping for the same GUID when using e.g. http://www.generalarcade.com/gamepadtool/ ?

If so, it might be that the mapping in https://github.com/gabomdq/SDL_GameControllerDB is wrong and it should be updated.

You can test the new mapping you got in Godot with:
```
SDL_GAMECONTROLLERCONFIG="your mapping here" ./path/to/godot
```

**Comment by skison** [comment]
Yes, it does appear that way. When I tried your command above with the newly generated mapping for that existing GUID, Godot interpreted it correctly.

Should I go ahead and open up an issue in the SDL_GameControllerDB repo?

**Comment by akien-mga** [comment]
> Should I go ahead and open up an issue in the SDL_GameControllerDB repo?

I'd say you can open a Pull Request directly on that repository with the updated mappings, on details on how you retrieved it and your gamepad. I'm not sure what their workflow is to decide if a mapping update is correct compared to the previous value, but they'll tell you.

**Comment by Calinou** [comment]
Does this occur if you install and use the [xpadneo](https://github.com/atar-axis/xpadneo) driver? There are many known quirks when using the Xbox One S controller in Bluetooth on Linux.

**Comment by skison** [comment]
@Calinou, I think you might be right about this being a driver issue. I was previously using xboxdrv when I reported these issues, but I just tried uninstalling/purging it and now things appear to be functioning as expected. The controller now shows up as "045e-028e-Microsoft X-Box 360 pad" on that gamepad test site, and simply as "Xinput Controller" on the SDL2 Gamepad Tool, and that same tool I used to generate mappings before now generates entirely different ones. I can start up Godot 3.1.2 without specifying a SDL_GAMECONTROLLERCONFIG string, and it still interprets the controls correctly.

Sorry for the misleading reports! I don't think this is an issue for Godot or SDL_GameControllerDB after all...
